feat(analytics): Implement analytics attributes and active state mana…

oidacra opened the pull request:
## Summary

  Enable analytics tracking in UVE by implementing active state management and React hooks. Content elements automatically receive analytics attributes when DotCMS Analytics initializes, while keeping UVE
  editing isolated from production tracking.

  ## Changes Made

  ### Analytics SDK (`core-web/libs/sdk/analytics/`)

  - **Configuration validation**: Added `validateAnalyticsConfig()` to check required `siteAuth` and `server` fields
  - **State management**: Introduced `window.__dotAnalyticsActive__` flag and `dotcms:analytics:ready` event
  - **Enhanced cleanup**: Reset analytics state and dispatch `dotcms:analytics:cleanup` event on teardown

  ### React SDK (`core-web/libs/sdk/react/`)

  - **New hook**: Created `useIsAnalyticsActive()` using `useSyncExternalStore` for reactive analytics state
  - **Contentlet updates**: Apply analytics attributes (`data-dot-analytics-*`) only when active and NOT in UVE mode
  - **Attribute separation**: UVE attributes only in dev mode, analytics attributes only in production

  ### UVE Core (`core-web/libs/sdk/uve/`)

  - **Utility function**: Added `isAnalyticsActive()` framework-agnostic checker
  - **Attribute builder**: Created `getDotAnalyticsAttributes()` to generate tracking attributes
  - **Type safety**: Added `DotAnalyticsAttributes` interface with guaranteed prefix
  - **Constants**: Centralized window keys (`ANALYTICS_WINDOWS_ACTIVE_KEY`, `ANALYTICS_WINDOWS_CLEANUP_KEY`)

  ### Tests

  - 35 new test cases across analytics, React components, and core utilities
  - Coverage: validation, state management, SSR handling, attribute application

  ## Technical Details

  **Event-driven architecture**: Uses custom DOM events (`dotcms:analytics:ready`, `dotcms:analytics:cleanup`) with `useSyncExternalStore` for optimal React 18+ performance. Works regardless of
  initialization order.

  **Attribute isolation**:
  - UVE mode: `data-dot-*` attributes only
  - Production + analytics active: `data-dot-analytics-*` attributes only
  - Mutual exclusivity prevents conflicts

  ## Breaking Changes

  None - backward-compatible enhancement.

  ## Testing

  - [x] Unit tests added (35 test cases)
  - [x] Manual testing performed
  - [ ] Integration/E2E tests not applicable


## Video
https://github.com/user-attachments/assets/8c5a5574-56b5-497d-b688-93201c34a1e4



  ## Related Issues

  Closes #33568

This PR fixes: #33568

oidacra committed:
feat(analytics): Implement analytics attributes and active state management

- Added constants for analytics window keys: ANALYTICS_WINDOWS_ACTIVE_KEY and ANALYTICS_WINDOWS_CLEANUP_KEY.
- Enhanced the initializeContentAnalytics function to manage the active state of analytics in the window object.
- Introduced a new hook, useIsAnalyticsActive, to monitor the analytics active state.
- Updated Contentlet component to conditionally add analytics attributes based on the active state.
- Implemented tests to validate the new analytics functionality and ensure proper behavior in different scenarios.

oidacra committed:
refactor(sdk-react): Simplify useIsAnalyticsActive hook and update documentation

oidacra committed:
fix(sdk-react): Update analytics attribute handling in development mode

oidacra committed:
feat(sdk-analytics): Enhance analytics initialization and state management

- Added dispatch of a custom event dotcms:analytics:ready in the initializeContentAnalytics function to notify when analytics is ready.
- Updated the useIsAnalyticsActive hook to listen for the dotcms:analytics:ready event, allowing components to reactively update when analytics becomes active.
- Improved test coverage for the new event handling in the analytics setup.

oidacra committed:
feat(sdk-analytics): Enhance cleanup process and state management

oidacra committed:
feat(sdk-analytics): Add validation for analytics configuration and enhance utility functions

oidacra committed:
feat(sdk-analytics): Refactor analytics window keys and enhance global interface

oidacra committed:
refactor(sdk-analytics): Update analytics key handling in tests and utility functions

zJaaal commented on core-web/libs/sdk/uve/src/lib/dom/dom.utils.ts:
```diff
@@ -283,6 +284,27 @@ export function getDotContentletAttributes(
+    contentlet: DotCMSBasicContentlet
+): DotAnalyticsAttributes {
+    return {
+        class: 'dotcms-analytics-contentlet',
```
do we need a class or a another attr? 

You can also query by attributes

oidacra replied:
Class-based queries are 3-5x faster than attribute-based queries when finding multiple elements:
```
// Fast path (~1-3ms for 1000 contentlets)
const allContentlets = document.querySelectorAll('.dotcms-analytics-contentlet');

// Slower path (~5-15ms for same query)
const allContentlets = document.querySelectorAll('[data-dot-analytics-identifier]');
```

Modern browser engines maintain optimized indexes for CSS classes
Data attributes, on the other hand, not use indexes

oidacra committed:
feat(sdk-analytics): Implement lazy initialization for analytics event listeners

oidacra committed:
feat(sdk-react): Add dynamic container class for analytics in Contentlet component

oidacra committed:
refactor(sdk-analytics): Clean up and reorganize analytics key imports and type definitions

- Removed duplicate import of ANALYTICS_WINDOWS_ACTIVE_KEY and ANALYTICS_WINDOWS_CLEANUP_KEY in dot-content-analytics.ts.
- Adjusted the DotAnalyticsAttributes type definition by removing the unnecessary class property.
- Improved import structure in core.spec.ts for better clarity and consistency.

oidacra committed:
refactor(sdk-analytics): Remove unnecessary whitespace and clean up test file imports
